sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query3.py

```diff
@@ -0,0 +1,160 @@
+      | nexmark_query_util.JustAuctions()
+      | 'query3_filter_category' >> beam.Filter(lambda auc: auc.category == 10)
+      | 'query3_key_by_seller' >> beam.ParDo(
+          nexmark_query_util.AuctionBySellerFn()))
+  person_by_id = (
+      windowed_events
+      | nexmark_query_util.JustPerson()
+      | 'query3_filter_region' >> beam.Filter(
+          lambda person: person.state == 'OR' or person.state == 'ID' or person.
+          state == 'CA')
+      | 'query3_key_by_person_id' >> beam.ParDo(
+          nexmark_query_util.PersonByIdFn()))
+  return ({
+      nexmark_query_util.AUCTION_TAG: auction_by_seller_id,
+      nexmark_query_util.PERSON_TAG: person_by_id,
+  }
+          | beam.CoGroupByKey()
+          | 'query3_join' >> beam.ParDo(
+              JoinFn(metadata.get('max_auction_waiting_time')))
+          | 'query3_output' >> beam.Map(
+              lambda t: {
+                  ResultNames.NAME: t[1].name,
+                  ResultNames.CITY: t[1].city,
+                  ResultNames.STATE: t[1].state,
+                  ResultNames.AUCTION_ID: t[0].id
+              }))
+
+
+class JoinFn(beam.DoFn):
+  """
+  Join auctions and person by person id and emit their product one pair at
+  a time.
+
+  We know a person may submit any number of auctions. Thus new person event
+  must have the person record stored in persistent state in order to match
+  future auctions by that person.
+
+  However we know that each auction is associated with at most one person, so
+  only need to store auction records in persistent state until we have seen the
+  corresponding person record. And of course may have already seen that record.
+  """
+
+  AUCTIONS = 'auctions_state'
+  PERSON = 'person_state'
+  PERSON_EXPIRING = 'person_state_expiring'
+
+  auction_spec = userstate.BagStateSpec(AUCTIONS, nexmark_model.Auction.CODER)
+  person_spec = userstate.ReadModifyWriteStateSpec(
+      PERSON, nexmark_model.Person.CODER)
+  person_timer_spec = userstate.TimerSpec(
+      PERSON_EXPIRING, userstate.TimeDomain.WATERMARK)
+
+  def __init__(self, max_auction_wait_time):
+    self.max_auction_wait_time = max_auction_wait_time
+
+  def process(
+      self,
+      element,
+      auction_state=beam.DoFn.StateParam(auction_spec),
+      person_state=beam.DoFn.StateParam(person_spec),
+      person_timer=beam.DoFn.TimerParam(person_timer_spec)):
+    # extract group with tags from element tuple
+    _, group = element
+
+    existing_person = person_state.read()
+    if existing_person:
+      # the person exists in person_state for this person id
+      for auction in group[nexmark_query_util.AUCTION_TAG]:
+        yield auction, existing_person
+      return
+
+    new_person = None
+    for person in group[nexmark_query_util.PERSON_TAG]:
+      if not new_person:
+        new_person = person
+      else:
+        logging.error(
+            'two new person wtih same key: %s and %s' %
+            (repr(person), repr(new_person)))
```

Comment:
'repr' seems redundant

Reply:
the person and new person here are actual person objects so I think we need to call repr to convert them to string